go/vt/vttablet/tabletserver/repltracker/writer.go

```diff
@@ -133,34 +163,52 @@ func (w *heartbeatWriter) Open() {
 	// block this thread, and we could end up in a deadlock.
 	// Instead, we try creating the database and table in each tick which runs in a go routine
 	// keeping us safe from hanging the main thread.
-	w.appPool.Open(w.env.Config().DB.AppWithDB())
-	w.allPrivsPool.Open(w.env.Config().DB.AllPrivsWithDB())
-	if w.onDemandDuration == 0 {
-		w.enableWrites(true)
-		// when onDemandDuration > 0 we only enable writes per request
-	} else {
+
+	if w.keyspaceShard != testKeyspaceShard {
+		// testKeyspaceShard indicates we're running in a unit test, in which case mock
+		// pools will have been opened.
+		w.appPool.Open(w.env.Config().DB.AppWithDB())
+		w.allPrivsPool.Open(w.env.Config().DB.AllPrivsWithDB())
+	}
+
+	if w.onDemandDuration > 0 {
+		// Clients are given access to RequestHeartbeats(). But such clients can also abuse
+		// the system by initiating heartbeats too frequently. We rate-limit these requests.
+		rateLimiter := timer.NewRateLimiter(w.onDemandDuration / 4)
+		atomic.StorePointer(w.unsafeOnDemandRequestsRateLimiter(), unsafe.Pointer(rateLimiter))
+	}
+
+	switch w.configType {
+	case HeartbeatConfigTypeNone:
+		// Do not kick any heartbeats
+		return
```

Comment:
So this is one practical difference currently between None and OnDemand. What's the significance of that (not doing any initial request)?

Reply:
Honestly the different is not significant. We can debate whether we at all need the initial request.